### DISCUSSION TITLE
Is there a Python syntax to write/export a model as mps file using Highspy?

### DISCUSSION THREAD

**Discussion by hbshrestha** [ERGO-Code/HiGHS#2059] [Q&A] [answered]
The HIGHS [documentation](https://ergo-code.github.io/HiGHS/dev/interfaces/python/example-py/#Build-a-model) provides a simple example of building  a model using highspy.
Is there an option to write or export mps file from the below given model using highspy?

```
import highspy
import numpy as np
h = highspy.Highs()

x0 = h.addVariable(lb = 0, ub = 4)
x1 = h.addVariable(lb = 1, ub = 7)

h.addConstr(5 <=   x0 + 2*x1 <= 15)
h.addConstr(6 <= 3*x0 + 2*x1)

h.minimize(x0 + x1)
```

**Comment by jajhall** [comment, marked as answer]
Yes, `h.writeModel("foo.mps")`

**Reply by hbshrestha** [reply]
Thank you for the response!

For the above model, I get the `<HighsStatus.kError: -1>` message.
In the terminal, I see `ERROR:   Model has repeated column names`. Not sure why this is the case.

**Reply by jajhall** [reply]
Did you define the variable names? If, so, could there be duplicates?

**Reply by hbshrestha** [reply]
The mps file is not exported due to the error message in my view.

**Reply by jajhall** [reply]
Indeed, if there are duplicate names, the MPS file won't be created. 

For me the question is why there are duplicate names. If HiGHS creates them, they should be guaranteed to be distinct, which is why I asked whether you define them

**Reply by hbshrestha** [reply]
I didn't define the variable names in the code above. I realized that the names have to specified manually as a separate parameter.
By assigning `name = "x0"` and `name = "x1"`, I could finally resolve this error and export mps file.

```
x0 = h.addVariable(lb=0, ub=4, name="x0")
x1 = h.addVariable(lb=1, ub=7, name="x1")
```

**Reply by jajhall** [reply]
Thanks. What this tells me is that the default empty name of "" is being stored for each name in HiGHS, so it considered there to be duplicate names. I need to check that empty names aren't stored in this way

**Reply by hbshrestha** [reply]
Indeed, that's correct.

**Reply by jajhall** [reply]
I've tried to duplicate this behaviour, and I get the message 

> WARNING: There are empty or excessively-long column names: using constructed names with prefix "c"
WARNING: There are empty or excessively-long row names: using constructed names with prefix "r"

This is because there are no names associated with columns or rows. If the name isn't specified in the call to `h.addVariable` then, for the default value `name: Optional[str] = None`, no (blank) name is passed to HiGHS.

**Reply by rschwarz** [reply]
As a somewhat off-topic comment on this method: It seems that by default, it produces MPS in "fixed format" and HiGHS will truncate the coefficients to 10 digits of precision:

https://github.com/ERGO-Code/HiGHS/blob/4a7f24ac63b3832e30694969e0a69a4b3c1f1674/src/io/HMPSIO.cpp#L792

I've found that in the C++ method, there is a boolean flag where one can switch to a free format (with full precision, and also longer names for rows and columns), but the Python wrapper does not yet provide this choice, currently.

**Reply by jajhall** [reply]
Strange. My understanding is that MPS files are written out in free format with full precision, particularly if the names are longer than 8 characters, and regardless of calling language, with the fixed format reader only used if row/column names contain spaces. I'll check, nonetheless

**Reply by rschwarz** [reply]
Yes, I think this would be the best and safest default.
